client/base_client.go

```diff
@@ -39,63 +39,119 @@ const (
 	memberUpdateInterval = time.Minute
 )
 
-// baseClient is a basic client for all other complex client.
-type baseClient struct {
-	urls      atomic.Value // Store as []string
-	clusterID uint64
+type tsoRequest struct {
+	start      time.Time
+	clientCtx  context.Context
+	requestCtx context.Context
+	done       chan error
+	physical   int64
+	logical    int64
+	dcLocation string
+	keyspaceID uint32
+}
+
+// BaseClient defines the general interface for service discovery on a quorum-based cluster
+// or a primary/secondy configured cluster.
+type BaseClient interface {
+	// Init initialize the concrete client underlying
+	Init() error
+	// Close all grpc client connnections
+	CloseClientConns()
+	// GetClusterID returns the ID of the cluster
+	GetClusterID(context.Context) uint64
+	// GetURLs returns the URLs of the servers.
+	GetURLs() []string
+	// GetTSOAllocators returns {dc-location -> TSO allocator leader URL} connection map
+	GetTSOAllocators() *sync.Map
+	// GetTSOAllocatorServingAddrByDCLocation returns the tso allocator of the given dcLocation
+	GetTSOAllocatorServingAddrByDCLocation(dcLocation string) (string, bool)
+	// GetTSOAllocatorClientConnByDCLocation returns the tso allocator grpc client connection
+	// of the given dcLocation
+	GetTSOAllocatorClientConnByDCLocation(dcLocation string) (*grpc.ClientConn, string)
+	// GetServingEndpointAddr returns the grpc client connection of the serving endpoint
+	// which is the leader in a quorum-based cluster or the primary in a primary/secondy
+	// configured cluster.
+	GetServingEndpointClientConn() *grpc.ClientConn
+	// GetServingEndpointAddr returns the serving endpoint which is the leader
+	// in a quorum-based cluster or the primary in a primary/secondy configured cluster.
+	GetServingEndpointAddr() string
+	// GetBackupEndpointsAddrs gets the addresses of the current reachable and healthy
+	// backup service endpoints randomly. Backup service endpoints are followers in a
+	// quorum-based cluster or secondaries in a primary/secondary configured cluster.
+	GetBackupEndpointsAddrs() []string
+	// GetOrCreateGRPCConn returns the corresponding grpc client connection of the given addr
+	GetOrCreateGRPCConn(addr string) (*grpc.ClientConn, error)
+	// ScheduleCheckIfMembershipChanged is used to trigger a check to see if there is any
+	// membership change among the leader/followers in a quorum-based cluster or among
+	// the primary/secondaries in a primary/secondy configured cluster.
+	ScheduleCheckIfMembershipChanged()
+	// Immediately checkif there is any membership change among the leader/followers in a
+	// quorum-based cluster or among the primary/secondaries in a primary/secondy configured cluster.
+	CheckIfMembershipChanged() error
```

Comment:
their name is so long, how about `ScheduleCheckMemberChanged` and `CheckMemberChanged`

Reply:
No. It's to schedule a check instead of checking immediatelly.

Reply:
I'll Change from ScheduleCheckIfMembershipChanged to ScheduleCheckMemberChanged

Reply:
changed.